### PR TITLE
ReplyTextView: Interactive Dismiss

### LINKED ISSUE #4358: ReplyTextView should be dismissed interactively
Let's update all of the classes that use the ReplyTextView component, so that it gets dismissed interactively, on swipe.

- [x] NotificationDetailsViewController
- [x] CommentViewController
- [x] ReaderCommentsViewController

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Details:
In this PR we're implementing `KeyboardDismissHelper`. The new kid on the block wraps up all of the logic related to:

- Updating a bottomConstraint, whenever the keyboard shows up
- Detecting the **interactiveDismiss** event, on the fly, and updating the bottom constraint as well

This has been implemented in the Reader, Notifications and Comments.

Fixes #4358

#### Scenario A: Notifications
1. Log into a dotcom account, using an iPhone device
2. Open a Notifications Y comment
3. Verify that the replyTextView is visible

Perform the Interactive Dismiss gesture. Play a bit with it: lift your finger before it's complete, and verify that it returns to place.

**Please:** Verify that the replyTextView doesn't show up on iPad Devices.

#### Scenario B: Comments
1. Log into a dotcom account, using an iPhone device
2. Open any of your Site's Details
3. Open the Comments tab
4. Open any of the site's comments
5. Verify that the replyTextView is visible

Perform the Interactive Dismiss gesture. Play a bit with it: lift your finger before it's complete, and verify that it returns to place.

**Please:** Verify that the replyTextView doesn't show up on iPad Devices.

#### Scenario C: Reader Comments
1. Log into a dotcom account
2. Open a post that contains comments (preferrably +10 comments!)
3. Open the Comments List
4. Verify that the replyTextView is visible (should be visible on both iPad / iPhone)
5. Perform the Interactive Dismiss gesture. Verify it doesn't break
6. Dismiss the keyboard.
7. Swipe over the comments table, so that the table begins scrolling
8. Tap over the replyTextField to show the keyboard, and verify nothing breaks
9. Try the interactive dismiss, and cancel the gesture at the middle.

Needs review: @aerych 
Sir, up for a fun review?

Thanks in advance!

### CODE REVIEW

File: WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    ///
+    var dismissableControl: UIView?
+
+    /// Reference to the bottom layout constraint
+    ///
+    var bottomLayoutConstraint: NSLayoutConstraint
+
+    /// Closure to be executed whenever the Keyboard will be Hidden
+    ///
+    var onWillHide: (Void -> ())?
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift]: Maybe its just me but `Void -> ()` seems odd.  I know its syntactically correct but it reminds me of ObjC where the return type is a prefix and arguments follow.  Would it be more swifty to reverse it? `() -> Void`?

File: WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+            let curve = UIViewAnimationCurve(rawValue: rawCurve) else
+        {
+            return
+        }
+
+        // Bottom Inset: Consider the tab bar!
+        let convertedKeyboardRect = parentView.convertRect(kbRect.CGRectValue(), fromView: nil)
+        let bottomInset = max(convertedKeyboardRect.height - convertedKeyboardRect.maxY + parentView.frame.height, 0)
+
+        UIView.beginAnimations(nil, context: nil)
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift]: Why not a block based animation?

**Comment by aerych** [comment]
On the iPad, if I hold the keyboard button and choose undock or split I can get some wonky layout. 

simulator screen shot jul 22 2016 8 56 29 am

simulator screen shot jul 22 2016 8 51 34 am

simulator screen shot jul 22 2016 8 51 43 am

File: WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift]: You're actually right!! `() -> Void` **IS** the right way. Thanks for pointing this out!

File: WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift]: Reason why is... the `animationCurve` parameter. Although the blocks API accepts the `UIViewAnimationOptions`, and the Curve can be converted... problem there si that there's a `delay` parameter.

Apparently, setting the delay to Zero still causes the animation to be fired with an *extremely small* delay, noticeable with Slow Animations enabled.

Furthermore, calling the `setAnimationCurve` function before running the plain block-based API seems to be disregarding the (statically-set) curve. I'm afraid that... the old school API is what brings us closer to the actual Keyboard animation.

(**Just** tested on iOS 9 simulator with Slow Animations enabled).

Thanks for pointing it out!!!

Comment (aerych) [inline WordPress/Classes/ViewRelated/Notifications/Tools/KeyboardDismissHelper.swift]: Makes sense.  Thanks for the detailed explanation!

**Comment by jleandroperez** [comment]
@aerych Fast Dismiss bug should be fixed now!. As per the Detached Keyboard, i'm afraid we can't really animate the replyView along (since none of the notifications gets posted live during the repositioning).

However, as soon as you lift the finger, the ReplyTextView should get repositioned properly.

Ready for another look!

And again, many many thanks for testing this!!!

**Comment by jleandroperez** [comment]
@aerych ready for another look, and (hopefully!!!) good to go perhaps!

Thanks again and over and over again!!!

**Comment by aerych** [comment]
Awesome sir! Looks great :)
:shipit: with great 🚢 ing

**Comment by jleandroperez** [comment]
Thaaaaanks @aerych !! 💯

**Pull request merged**